internal/journal/omap.go

```diff
@@ -0,0 +1,138 @@
+package journal
+
+import (
+	"context"
+
+	"github.com/ceph/ceph-csi/internal/util"
+
+	"github.com/ceph/go-ceph/rados"
+	"k8s.io/klog"
+)
+
+// listExcess is the number of false-positive key-value pairs we will
+// accept from ceph when getting omap values.
+var listExcess = 32
+
+func getOMapValues(
+	ctx context.Context,
+	conn *Connection,
+	poolName, namespace, oid, prefix string, keys []string) (map[string]string, error) {
+	// fetch and configure the rados ioctx
+	ioctx, err := conn.conn.GetIoctx(poolName)
+	if err != nil {
+		return nil, omapPoolError(poolName, err)
+	}
+	defer ioctx.Destroy()
+
+	if namespace != "" {
+		ioctx.SetNamespace(namespace)
+	}
+
+	results := map[string]string{}
+	// want is our "lookup map" that ensures O(1) checks for keys
+	// while iterating, without needing to complicate the caller.
+	want := make(map[string]bool, len(keys))
+	for i := range keys {
+		want[keys[i]] = true
+	}
+
+	err = ioctx.ListOmapValues(
+		oid, "", prefix, int64(len(want)+listExcess),
+		func(key string, value []byte) {
+			if want[key] {
+				results[key] = string(value)
+			}
+		},
+	)
+	switch err {
+	case nil:
+	case rados.ErrNotFound:
+		klog.Errorf(
+			util.Log(ctx, "omap not found (pool=%q, namespace=%q, name=%q): %v"),
+			poolName, namespace, oid, err)
+		return nil, util.NewErrKeyNotFound(oid, err)
+	default:
+		return nil, err
+	}
+
+	klog.V(4).Infof(
+		util.Log(ctx, "got omap values: (pool=%q, namespace=%q, name=%q): %+v"),
+		poolName, namespace, oid, results)
+	return results, nil
+}
+
+func removeMapKeys(
+	ctx context.Context,
+	conn *Connection,
+	poolName, namespace, oid string, keys []string) error {
+	// fetch and configure the rados ioctx
+	ioctx, err := conn.conn.GetIoctx(poolName)
+	if err != nil {
+		return omapPoolError(poolName, err)
+	}
+	defer ioctx.Destroy()
+
+	if namespace != "" {
+		ioctx.SetNamespace(namespace)
+	}
+
+	err = ioctx.RmOmapKeys(oid, keys)
+	switch err {
+	case nil:
+		klog.V(4).Infof(
```

Comment:
can we move this logging to the end of the function? similar ways as its done in getOMapValues